### DISCUSSION TITLE
document.deleteOne() does not trigger the post hook of deleteOne middleware

### DISCUSSION THREAD

**Discussion by mrdulin** [Automattic/mongoose#13624] [Q&A] [answered]
From the [middleware](https://mongoosejs.com/docs/middleware.html) documentation:

> Note: Unlike schema.pre('remove'), Mongoose registers updateOne and deleteOne middleware on Query#updateOne() and Query#deleteOne() by default. **This means that both `doc.updateOne()` and `Model.updateOne()` trigger updateOne hooks**, but this refers to a query, not a document. To register updateOne or deleteOne middleware as document middleware, use schema.pre('updateOne', { document: true, query: false }).

So I think `doc.deleteOne()` and `Model.deleteOne()` will trigger the `deleteOne` hook. But `doc.deleteOne()` does not trigger the `deleteOne` hook.

For example:

```ts
import mongoose from 'mongoose';
import { config } from '../../config';

mongoose.set('debug', true);
console.log(mongoose.version);

const reviewSchema = new mongoose.Schema({});
reviewSchema.post('deleteOne', function () {
	console.log('post deleteOne');
});
reviewSchema.post('updateOne', function () {
	console.log('post updateOne');
});
const Review = mongoose.model('Review', reviewSchema);

(async function main() {
	try {
		await mongoose.connect(config.MONGODB_URI);
		// seed
		const [r1] = await Review.create([{}, {}]);
		// test
		// updateOne
		await r1?.updateOne({ name: 'teresa teng' }); // works
		await Review.updateOne({ _id: r1?._id }, { name: 'teresa teng' }); // works

		// deleteOne
		await Review.deleteOne({ _id: r1?._id }); // works
		await r1?.deleteOne(); // doesn't work
	} catch (error) {
		console.error(error);
	} finally {
		await mongoose.connection.close();
	}
})();
```

Logs:
```zsh
7.4.0
Mongoose: reviews.insertOne({ _id: ObjectId("64b771cc02ca30b633518b8e"), __v: 0 }, {})
Mongoose: reviews.insertOne({ _id: ObjectId("64b771cc02ca30b633518b8f"), __v: 0 }, {})
post updateOne
post updateOne
Mongoose: reviews.deleteOne({ _id: ObjectId("64b771cc02ca30b633518b8e") }, {})
post deleteOne
Mongoose: reviews.deleteOne({ _id: ObjectId("64b771cc02ca30b633518b8e") }, { session: null })
```

Is this a bug or am I missing something?

**Comment by IslandRhythms** [comment]
change your post function to 
```javascript
reviewSchema.post('deleteOne', {document: true, query: true}, function () {
	console.log('post deleteOne');
});
```

**Reply by mrdulin** [reply]
My question is why the default behavior of the `updateOne()` and `deleteOne()` hooks are not the same. Is this by design? Besides, the documentation is not clear.

**Reply by vkarpov15** [reply, marked as answer]
Unfortunately the default behavior is slightly different between `updateOne` and `deleteOne`, and we have tests that assert on this behavior. But we will change this behavior in 8.x, see #13660, specifically this test file: https://github.com/Automattic/mongoose/pull/13660/files#diff-7b7bf89c426b1b002315ff19e892ac3a7750e57ed11ad563074b4e200fad218e. Will update our docs.
